packages/gatsby/src/query/queue.js

```diff
@@ -100,15 +92,22 @@ const processBatch = async (queue, jobs, activity) => {
     return Promise.resolve()
   }
 
-  const runningJobs = jobs.map(job =>
-    pushJob(queue, job).then(v => {
-      if (activity.tick) {
-        activity.tick()
-      }
-      return v
-    })
-  )
-  return Promise.all(runningJobs)
+  return new Promise((resolve, reject) => {
+    if (activity.tick) {
+      queue.on(`task_finish`, () => activity.tick())
+    }
+
+    queue
+      .on(`task_failed`, (...err) => {
+        // Note: the first arg is the path, the second the error
+        reject(err)
+        return err
+      })
+      // Note: `drain` fires when all tasks _finish_, `empty` fires when queue is empty (but tasks are still running)
+      .on(`drain`, resolve)
```

Comment:
Should we clean up listeners at some point? Maybe use `once` for both `task_failed` and `drain` instead of `on`?

Reply:
This queue is created for this task and it will be gc'ed once it's released in processQueries (in query/index.js). So I don't think we need to tear the listeners down.

I'm fine with using `once` instead of `on`. Shouldn't make a difference but is cleaner.